### ISSUE TITLE
haproxy: CVE-2017-3738: OpenSSL rsaz_1024_mul_avx2 overflow on x86_64 (SecAdv 20171207)

### ISSUE THREAD

**Issue by zvercodebender** [rancher/rancher#12064]
An overflow bug in the AVX2 Montgomery multiplication procedure exists in the detected version of OpenSSL. This bug does not affect EC (elliptic curve) based algorithms. According to the vendor, attacks against RSA and DSA are unlikely, and against the Diffie-Hellman (with 1024-bit moduli) are considered feasible, however very complex (resource consuming). For the attack, to be successful, the server would have to share the same DH1024 private key between sessions with different clients which is not the case in modern OpenSSL versions (after the CVE-2016-0701 has been fixed in OpenSSL 1.0.2f).

Furthermore, this bug affects only CPUs that support Intel AVX2 extension, but do not support ADX extension (ex. 4th Generation Intel CPUs "Haswel").

NOTE: This bug is considered to be LOW severity, and the vendor did not made a new release in 1.1.0 branch to fix it. The bug in 1.1.0 branch will be fixed in 1.1.1h (when this version is released).

**Remediation**: Upgrade OpenSSL to the latest available version. This particular vulnerability has been addressed in versions: 1.0.2n (for 1.0.2 branch), and 1.1.0h (for 1.1.0 branch).

https://www.openssl.org/news/secadv/20171207.txt

---
| Useful | Info |
| :-- | :-- |
|Versions|Rancher `v1.6.14` Cattle: `v0.183.37` UI: `v1.6.37` |
|Access|`openldap` `admin`|
|Orchestration|`Cattle`|
|Route|`stacks.index`|
| openssl | OpenSSL 1.0.2g  1 Mar 2016 |

**Comment by deniseschannon** [comment]
@superseb can you test with `lb-service-haproxy:v0.9.0`?

**Comment by superseb** [comment]
```
Ubuntu 16.04 LTS (Xenial Xerus): | released (1.0.2g-1ubuntu4.10)
```
Source: https://people.canonical.com/~ubuntu-security/cve/2017/CVE-2017-3738.html

```
$ docker run -t --entrypoint bash rancher/lb-service-haproxy:v0.9.0 -c "dpkg -l | grep openssl"
ii  openssl                     1.0.2g-1ubuntu4.10                    amd64        Secure Sockets Layer toolkit - cryptographic utility
```

@zvercodebender Let me know if you don't agree.

**Comment by zvercodebender** [comment]
The issue is the 1.0.2g has the problem.  Upgrading to 1.0.2n or 1.1.0h will solve the problem unless Ubuntu 16.0 LTS has back ported those problems.  After further review it might be that the Ubuntu distribution is using a back ported version of the package based on the changelog

NAME="Ubuntu"
VERSION="16.04.4 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.4 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

Get:1 http://changelogs.ubuntu.com openssl 1.0.2g-1ubuntu4.10 Changelog [155 kB]
Fetched 155 kB in 0s (218 kB/s)
openssl (1.0.2g-1ubuntu4.10) xenial-security; urgency=medium

  * SECURITY UPDATE: Read/write after SSL object in error state
    - debian/patches/CVE-2017-3737-pre.patch: add test/ssltestlib.*,
      add to test/Makefile.
    - debian/patches/CVE-2017-3737-1.patch: don't allow read/write after
      fatal error in ssl/ssl.h.
    - debian/patches/CVE-2017-3737-2.patch: add test to ssl/Makefile,
      ssl/fatalerrtest.c, test/Makefile.
    - CVE-2017-3737
  * SECURITY UPDATE: rsaz_1024_mul_avx2 overflow bug on x86_64
    - debian/patches/CVE-2017-3738.patch: fix digit correction bug in
      crypto/bn/asm/rsaz-avx2.pl.
    - CVE-2017-3738
